### ISSUE TITLE
URL path with colon (":") results in 404 not found

### ISSUE THREAD

**Issue by dwinston** [Kludex/starlette#1674]
### Discussed in https://github.com/encode/starlette/discussions/1673

<div type='discussions-op-text'>

<sup>Originally posted by **dwinston** June  3, 2022</sup>
I find that I can't GET or POST to a path containing a colon (":"). An example in FastAPI (which uses Starlette):

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/items:queue")
def queue_new_item():
    return "An item has been queued for later inclusion."
```

On `uvicorn main:app --reload` (FYI Python 3.9.12 and FastAPI 0.78.0 on MacOS), I get a 404 response from `GET http://127.0.0.1:8000/items:queue`.

I also filed this issue in the fastapi repo (https://github.com/tiangolo/fastapi/issues/4984), but I suspect that Starlette's use of colons for path parameter converters (https://www.starlette.io/routing/) may be playing a role here.

Is this a bug? I am trying to follow the official google cloud APIs "custom methods" pattern <https://cloud.google.com/apis/design/custom_methods>:

>For custom methods, they should use the following generic HTTP mapping:
> https://service.name/v1/some/resource/name:customVerb
</div>

**Comment by dwinston** [comment]
```python
from starlette.routing import compile_path

compile_path("/queries:run")
```
outputs
```
(re.compile(r'^/queries$', re.UNICODE), '/queries:run', {})
```
This seems wrong!

**Comment by dwinston** [comment]
I suspect https://github.com/encode/starlette/commit/7c7ec5a7f72de360bafa938d14e2e1d6f4b6cb69 broke this. It works in starlette 0.17.1 (I have a deployment using this version; I was wondering why colon-including routes are not working for me locally, where I am using 0.19.1  -- via latest FastAPI: https://fastapi.tiangolo.com/release-notes/), and I see that this commit was included since 0.18.0.

/cc @WhiteApfel @aminalaee @Kludex

**Comment by Kludex** [comment]
- Closed by https://github.com/encode/starlette/pull/1675

The fix will be available on Starlette 0.20.2.
